oeps/oep-0028-content-theming-in-react.rst

```diff
@@ -0,0 +1,246 @@
+
+Decision
+--------
+
+* We will build small modular components which will be non-customizable, stateless and contain HTML, and certain logic. We can see an example of a smaller component SiteLogo, which will later be used in a customizable component _Header.
+
+.. code-block:: js
+
+    const SiteLogo = (props) => {
+        return (
+            <img src={props.url} alt={props.altText} />
+        );
+    }
+
+* We will build the complete UI out of the above small modular components. We will generally use composition to build a customizable component. However, inheritance can be used at times. These customizable components will neither contain HTML nor logic. We can see an example of a customizable component called the _Header component, which is formed by the composition of SiteLogo, MainNav and UserAvatar component.
+
+.. code-block:: js
+
+    class _Header extends React.PureComponent {
+        render() {
+            return (
+                <header>
+                    <SiteLogo/>
+                    <MainNav/>
+                    <UserAvatar/>
+                </header>
+            );
+        }
+    }
+
+* We will follow a 2 step approach to create any UI Component. We will first break down a complex component into smaller non-customizable components. We will then compose these smaller components to form a complex UI.
+
+We can see this through an example where we have a theme, which has several components for a default theme.
+
+.. code-block:: js
+
+    // Customizable Header
+    class _Header extends React.PureComponent {
+        render() {
+            return (
+                <header>
+                    <SiteLogo/>
+                    <MainNav/>
+                    <UserAvatar/>
+                </header>
+            );
+        }
+    }
+
+    // Customizable Main Navigation Area
+    class _MainNav extends React.PureComponent {
+        render() {
+            return (
+                <MainNavWrapper>
+                    <a href="/">Home</a>
+                    <LoginLink/>
+                    {this.extraNavLinks}
+                </MainNavWrapper>
+            );
+        }
+        get extraNavLinks() { return []; }
+    }
+    // Internal MainNavWrapper - not meant to be modified in most cases
+    class _MainNavWrapper extends React.PureComponent {
+        render() {
+            return (
+                <div className="mainNav">
+                    <ul>
+                        {React.Children.map(this.props.children, (child) => (child ? <li>{child}</li> : null))}
+                    </ul>
+                </div>
+            )
+        }
+    }
+
+    // Default Theme:
+    export const Header = _Header;
+    export const MainNav = _MainNav;
+    export const MainNavWrapper = _MainNavWrapper;
+
+
+Now if we want to customize our _Header component, we can easily do it like
+
+.. code-block:: js
+
+    // SiteLogo being updated in Header
+    class MyThemedHeader extends _Header {
```

Comment:
As @bradenmacdonald mentioned elsewhere, this is using inheritance not composition.

Reply:
Perhaps I think you are talking about [this](https://github.com/taranjeet/open-edx-proposals/pull/1#discussion_r212421128) comment. That was for another point wherein the point I talked about using "composition" but gave an example of "composition". I updated it then only.

Also here it makes sense to use "inheritance" only since if we use composition we would end up creating components like 'BaseHeader`, `Header`, `SiteLogoHeader` and `MyCustomAnimatedLogoWidgetHeader`.

Let me know if this is not right.

Reply:
My point is that you mention in the paragraph: "We will prefer composition whenever adding any new component or content and inheritance when replacing or modifying the component." Immediately after that, you are showing an example of using inheritance when that is not the preferred approach. Show both ways of doing this, so that it's clearer which is the preferred approach in what circumstance. 

I would also say this is not a good example of inheritance at all. You are extending an existing component, but there is nothing from the original component that you are actually using. Why couldn't this just be a new component? The only thing it gets from ``_Header`` is the ``getNavLinks`` getter method and that isn't even being used, the ``render`` method is entirely replaced. 

Also, I think if a component has a good chance of needing modification in some area, there should probably be a hook for it. So, in this case, replacing the Logo in the ``_Header`` is a pretty obvious modification, so perhaps there should already be a prop for it, and should just default to the standard logo.

Reply:
@xitij2000 : I am not sure here but I think we should skip this difference part between composition and inheritance since that is something more specific to React and not this proposal in general. The line where we can say that here inheritance or composition which one should be used is very blur.

Keeping this in mind, I think its better to just write `via composition/inheritance` and leave it to the one who is implementing it because he/she will be a better judge of the situation. 

I have also commented on the ticket about how we can use composition for the Header component but that turned out to be a discussion on inheritance vs composition.

```
class Header extends React.PureComponent {
    public render() {
        return (
            <header>
                {props.children}
            </header>
        );
    }
}
 
const BaseHeader = (props) => {
    <Header>
        {props.children}
        <MainNav/>
        <UserAvatar/>
    </Header>
}
const SiteLogoHeader = (props) => {
    <BaseHeader>
        <SiteLogo />
    </BaseHeader>
}
 
const MyCustomAnimatedLogoHeader = (props) => {
    <BaseHeader>
        <MyCustomAnimatedLogo />
    </BaseHeader>
}

``` 
I think Braden can help us here.
@bradenmacdonald : Your inputs please :)

Reply:
I believe that inheritance will be necessary any time you want to _remove_ or _replace_ an existing component child such as the `SiteLogo` in the example above, unless there is some other hook for doing so via props (like `logoOverride=<MyCustomAnimatedLogo />`) or modifying a global (like replacing a component globally via `const SiteLogo = MyCustomSiteLogo; // not StandardSiteLogo;`)

Reply:
@bradenmacdonald How will inheritance help in that case though? I might be missing something but here is how I see it:
Case: You have a component where the logo should be removable/replaceable. 
* Provide a prop that allows replacing the logo. e.g. ``<div> {this.props.logo === undefined?<DefaultLogo />: this.props.logo}</div>``. This allows setting the logo as null or providing a custom logo without inheritance. 
* You inherit from the original component. But now you're overriding the original component's render method. So what exactly do you gain by inheriting vs creating a new component? The use for inheritance would be with components that have a lot of logic that will be inherited, but customisable components won't have that.

Reply:
> Maybe we should come up with more concrete examples to figure this out better? It's still fairly theoretical at this point.

Yup, I'm just having trouble seeing a component that has a such a straightforward render function, but it still makes sense to inherit, vs create a new component. I can see the case if there is some lifecycle code or other kind of complex code in the component, but I assumed that customisable components wouldn't have that.